model-optimizer/extensions/middle/ConcatOptimization.py

```diff
@@ -98,22 +94,39 @@ def find_and_replace_pattern(self, graph: Graph):
                     p_id += 1
 
 
-class ConcatOdInputEraser(MiddleReplacementPattern):
+class ConcatOdInputEraserAndPortsReconnect(MiddleReplacementPattern):
     """
-    Disconnects empty inputs of Concat operations -- as there is nothing to concatenate
+    The transformation performs two actions with Concat operations:
+    1. Disconnects empty inputs (input tensor has at least one input dimension equal to 0)
+    2. Renumber Concat inputs to be 0, 1, 2,...
     """
     enabled = True
     force_clean_up = True
 
     def find_and_replace_pattern(self, graph: Graph):
         for concat in graph.get_op_nodes(type='Concat'):
+            num_connected_ports = len([port for port in concat.in_ports().values() if not port.disconnected()])
             for in_port in concat.in_ports().values():
                 if in_port.disconnected():
                     continue
                 shape = in_port.data.get_shape()
                 assert shape is not None
-                if np.array_equal(shape, [0]):
+                if 0 in shape:
                     in_port.disconnect()
 
-            connected_input_ports = [in_port for in_port in concat.in_ports().values() if not in_port.disconnected()]
-            assert len(connected_input_ports), 'Concat {} does nothing'.format(concat.soft_get('name', concat.id))
+            connected_ports = [port for port_idx, port in sorted(concat.in_ports().items()) if not port.disconnected()]
+            assert len(connected_ports), 'Concat "{}" have no inputs after removing inputs with 0 dimensions' \
+                                         ''.format(concat.soft_get('name', concat.id))
+
+            max_port_id = max([port.idx for port in connected_ports])
+            # if we renumbered input ports or removed some of them we need to re-connect the inputs
+            if [port.idx for port in connected_ports] != list(range(len(connected_ports))) or \
+                    num_connected_ports != len(connected_ports):
```

Comment:
Why do we need the second check? First one covers all cases as I understand.

Reply:
No, it does not. Refer to test case `test_deletion_3`